src/libsyntax/parse/mod.rs

```diff
@@ -230,37 +239,52 @@ pub fn stream_to_parser(sess: &ParseSess, stream: TokenStream) -> Parser {
 /// Rather than just accepting/rejecting a given literal, unescapes it as
 /// well. Can take any slice prefixed by a character escape. Returns the
 /// character and the number of characters consumed.
-pub fn char_lit(lit: &str) -> (char, isize) {
+pub fn char_lit(lit: &str, diag: Option<(Span, &Handler)>) -> (Option<char>, isize) {
     use std::char;
 
     // Handle non-escaped chars first.
     if lit.as_bytes()[0] != b'\\' {
         // If the first byte isn't '\\' it might part of a multi-byte char, so
         // get the char with chars().
         let c = lit.chars().next().unwrap();
-        return (c, 1);
+        return (Some(c), 1);
     }
 
     // Handle escaped chars.
     match lit.as_bytes()[1] as char {
-        '"' => ('"', 2),
-        'n' => ('\n', 2),
-        'r' => ('\r', 2),
-        't' => ('\t', 2),
-        '\\' => ('\\', 2),
-        '\'' => ('\'', 2),
-        '0' => ('\0', 2),
+        '"' => (Some('"'), 2),
+        'n' => (Some('\n'), 2),
+        'r' => (Some('\r'), 2),
+        't' => (Some('\t'), 2),
+        '\\' => (Some('\\'), 2),
+        '\'' => (Some('\''), 2),
+        '0' => (Some('\0'), 2),
         'x' => {
             let v = u32::from_str_radix(&lit[2..4], 16).unwrap();
             let c = char::from_u32(v).unwrap();
-            (c, 4)
+            (Some(c), 4)
         }
         'u' => {
+            fn parse_unicode_escape(lit: &str, diag: Option<(Span, &Handler)>) -> Option<char> {
+                let s = lit.chars().filter(|&c| c != '_').collect::<String>();
+                let v = u32::from_str_radix(&s, 16).unwrap();
+                match char::from_u32(v) {
+                    Some(c) => Some(c),
+                    None => {
+                        err!(diag, |span, diag| {
+                            diag.struct_span_err(span, "invalid unicode character escape")
+                                .help("unicode escape must be at most 10FFFF")
+                                .emit();
+                                None
```

Comment:
Thanks for the review @petrochenkov!
Ok, I will do another round of changes and squash everything.